core/src/main/java/io/questdb/cutlass/text/TextImportTask.java

```diff
@@ -0,0 +1,995 @@
+            try (Path path = new Path().of(configuration.getInputRoot()).concat(inputFileName);
+                 DirectCharSink utf8Sink = new DirectCharSink(utf8SinkSize)) {
+                fd = ff.openRO(path.$());
+
+                TextLexer.Listener listener = (long line, ObjList<DirectByteCharSequence> fields, int fieldCount) ->
+                        this.onFieldsPartitioned(fields, fieldCount, utf8Sink);
+
+                long MASK = ~((255L) << 56 | (255L) << 48);
+                final long count = size / (2 * Long.BYTES);
+                long bytesToRead;
+                int additionalLines;
+
+                for (long i = 0; i < count; i++) {
+                    long lengthAndOffset = Unsafe.getUnsafe().getLong(address + i * 2L * Long.BYTES + Long.BYTES);
+                    long lineLength = lengthAndOffset >>> 48;
+                    offset = lengthAndOffset & MASK;
+                    bytesToRead = lineLength;
+                    additionalLines = 0;
+
+                    for (long j = i + 1; j < count; j++) {
+                        long nextLengthAndOffset = Unsafe.getUnsafe().getLong(address + j * 2L * Long.BYTES + Long.BYTES);
+                        long nextLineLength = nextLengthAndOffset >>> 48;
+                        long nextOffset = nextLengthAndOffset & MASK;
+
+                        //line indexing stops on first EOL char, e.g. \r, but it could be followed by \n
+                        long diff = nextOffset - offset - bytesToRead;
+                        if (diff > -1 && diff < 2 && bytesToRead + nextLineLength < bufLength) {
+                            bytesToRead += diff + nextLineLength;
+                            additionalLines++;
+                        } else {
+                            break;
+                        }
+                    }
+
+                    i += additionalLines;
+
+                    long n = ff.read(fd, buf, bytesToRead, offset);
+                    if (n > 0) {
+                        lexer.parse(buf, buf + n, Integer.MAX_VALUE, listener);
+                    } else {
+                        throw TextException.$("Can't read from file path='").put(path).put("', errno=").put(ff.errno()).put(",offset=").put(offset);
+                    }
+                }
+
+                lexer.parseLast();
+            } finally {
+                if (fd > -1) {
+                    ff.close(fd);
+                }
+
+                Unsafe.free(buf, bufLength, MemoryTag.NATIVE_DEFAULT);
+            }
+        }
+
+        private void mergePartitionIndexAndImportData(final FilesFacade ff,
+                                                      final Path partitionPath,
+                                                      final TextLexer lexer) throws TextException {
+            try (DirectLongList mergeIndexes = new DirectLongList(64, MemoryTag.NATIVE_DEFAULT)) {
```

Comment:
hardcoded capacity, this list can be "thread-local", and passed in. Typical setup is one Job instance per worker thread. This job instance can hold items like "path" and "list"

Reply:
Fixed.